### PR TITLE
WebGPURenderer: Shadows Node-Based

### PR DESCRIPTION
**Pull request by sunag**
**Description**

I made some modifications to the API so that we could do `RTT` or `Compute` at the time of rendering, thus creating a render tree if necessary, more precisely at Node update time, the implementation of which is clearer in the use of `Node.updateBefore()`, which is called in the process of the `WebGPURenderer.renderObject()` but before updating the matrices/uniforms. As the shadow is based on the architecture of the Nodes, it is not necessary to modify the core if you want to add a new filter or add a new type of shadow.

https://raw.githack.com/sunag/three.js/dev-webgpu-shadows/examples/webgpu_shadowmap.html

image

### CODE REVIEW

File: examples/jsm/nodes/core/NodeFrame.js [hunk 1]
```diff
@@ -34,6 +72,17 @@ class NodeFrame {
 
 			}
 
+		} else if ( updateType === NodeUpdateType.RENDER ) {
+
+			if ( this.rendererMap.get( node ) !== this.renderId || this.updateMap.get( node ) !== this.frameId ) {
```
Comment (LeviPesin) [inline examples/jsm/nodes/core/NodeFrame.js:77]: Why do we need a separate map? Maybe we can use one (and maybe rename it to `frameMap`)?

File: examples/jsm/nodes/lighting/AnalyticLightNode.js [hunk 2]
```diff
@@ -25,11 +37,102 @@ class AnalyticLightNode extends LightingNode {
...
 	}
 
-	update( /*frame*/ ) {
+	constructShadow( builder ) {
+
+		let shadowNode = this.shadowNode;
+
+		if ( shadowNode === null ) {
+
+			depthMaterial !== null || ( depthMaterial = new MeshBasicNodeMaterial() );
```
Comment (LeviPesin) [inline examples/jsm/nodes/lighting/AnalyticLightNode.js:46]:
```suggestion
			if ( depthMaterial === null ) depthMaterial = new MeshBasicNodeMaterial();
```

File: examples/jsm/nodes/core/NodeFrame.js (hunk 1 above)
Comment (sunag) [reply examples/jsm/nodes/core/NodeFrame.js:77]: `renderMap` would not be a `frameMap`, we can have several renderings, that is several RTT in a single frame, but we can rename, `frameMap` seems like a better name

**Comment by mrdoob** [comment]
Looking good! 🔥

**Pull request merged**